**SA463A &#x25aa; Data Wrangling and Visualization &#x25aa; Fall 2021 &#x25aa; Uhan**

# Lesson 15. Creating New Variables in Pandas

## In this lesson...

- More ways to create new variables in Pandas

<hr style="border-top: 2px solid gray; margin-top: 1px; margin-bottom: 1px"></hr>

## A tiny example dataset

* Let's start by importing Pandas:

In [1]:
import pandas as pd

* For this lesson, we'll use the example dataset in `data/sample.csv` to illustrate how the different ways of creating new variables work

In [2]:
df = pd.read_csv('data/sample.csv')
df

a         b   c          d
0  5  8.389033  10  14.115655
1  3  9.586854  17  15.843354
2  2  5.110890  17  15.194550
3  7  7.578774  14  15.297023
4  4  1.602499  18  14.786111

<hr style="border-top: 2px solid gray; margin-top: 1px; margin-bottom: 1px"></hr>

## The assign method

* In a previous lesson, we learned that we can create a new variable/column in a DataFrame by *direct assignment*, like this:

In [3]:
df['e'] = df['a'] + df['b']
df.head()

a         b   c          d          e
0  5  8.389033  10  14.115655  13.389033
1  3  9.586854  17  15.843354  12.586854
2  2  5.110890  17  15.194550   7.110890
3  7  7.578774  14  15.297023  14.578774
4  4  1.602499  18  14.786111   5.602499

* Note that the direct assignment approach makes the changes *in place*; that is, `df` itself is modified

- Here's another way to create a new variable/column in a DataFrame


- Let's read in the data again and start from scratch:

In [4]:
df = pd.read_csv('data/sample.csv')

- First, we create a function that takes a DataFrame as input, and outputs the sum of columns `a` and `b` of the DataFrame, like this:

In [5]:
# Solution
def add_a_and_b(x):
    '''
    Adds columns 'a' and 'b' in the input DataFrame x
    '''
    return x['a'] + x['b']

- Second, we use the `.assign()` DataFrame method to create the new variable
    - Each keyword argument `key=function` creates a new variable called `key` equal to the output of `function(df)`


- So, to create a new variable `e` that is equal to the sum of columns `a` and `b`, we can do this:

In [6]:
# Solution
df.assign(
    e=add_a_and_b
)

a         b   c          d          e
0  5  8.389033  10  14.115655  13.389033
1  3  9.586854  17  15.843354  12.586854
2  2  5.110890  17  15.194550   7.110890
3  7  7.578774  14  15.297023  14.578774
4  4  1.602499  18  14.786111   5.602499

* Note that `.assign()` *returns a new DataFrame*
    - It does not make the changes in place, like the direct assignment approach above


* If we inspect the contents of `df` now, we see that it doesn't contain a column named `e`:

In [7]:
df

a         b   c          d
0  5  8.389033  10  14.115655
1  3  9.586854  17  15.843354
2  2  5.110890  17  15.194550
3  7  7.578774  14  15.297023
4  4  1.602499  18  14.786111

* Now... creating a separate function to perform the operation we want is a bit cumbersome


* Instead, we can use a **lambda function**, like this:

In [8]:
# Solution
df.assign(
    e=lambda x: x['a'] + x['b']
)

a         b   c          d          e
0  5  8.389033  10  14.115655  13.389033
1  3  9.586854  17  15.843354  12.586854
2  2  5.110890  17  15.194550   7.110890
3  7  7.578774  14  15.297023  14.578774
4  4  1.602499  18  14.786111   5.602499

- The lambda function

    ```python
    lambda x: x['a'] + x['b']
    ```

    <br>takes `x` as input, and outputs `x['a'] + x['b']`, just like we defined in the function `add_a_and_b()`


- We can even create multiple variables in a single call of `.assign()`, like this:

In [9]:
# Solution
df.assign(
    e=lambda x: x['a'] + x['b'],
    f=lambda x: x['c'] * x['e']
)

a         b   c          d          e           f
0  5  8.389033  10  14.115655  13.389033  133.890326
1  3  9.586854  17  15.843354  12.586854  213.976518
2  2  5.110890  17  15.194550   7.110890  120.885130
3  7  7.578774  14  15.297023  14.578774  204.102841
4  4  1.602499  18  14.786111   5.602499  100.844988

- In the example above, note that we can even refer to *newly created* columns within the same call of `.assign()`

- **Template for using the `.assign()` method**

    ```python
    df.assign(
        new_variable_1=lambda x: ...,
        new_variable_2=lambda x: ...
    )
    ```

    <br>where `x` refers to the DataFrame `df`

### Why use the assign method?

- The direct assignment approach makes changes to the DataFrame in place, while `.assign()` returns a new DataFrame


- As a result, the `.assign()` method supports method chaining, like this:

In [10]:
(
    df
    .assign(e=lambda x: x['c'] / 2)
    .query('e > 5')
)

a         b   c          d    e
1  3  9.586854  17  15.843354  8.5
2  2  5.110890  17  15.194550  8.5
3  7  7.578774  14  15.297023  7.0
4  4  1.602499  18  14.786111  9.0

- *Refactoring* &mdash; in particular, renaming variables &mdash; is easier as well 


- As a comparison, think about changing the name of the DataFrame from `df` to `another_df` in these 2 cases:

    ```python
    # Case 1
    df['f'] = df['a'] + df['b'] + df['c'] + df['d']
    
    # Case 2
    df.assign(
        f=lambda x: x['a'] + x['b'] + x['c'] + x['d']
    )
    ```

<hr style="border-top: 2px solid gray; margin-top: 1px; margin-bottom: 1px"></hr>

## Useful Series and DataFrame methods for computations

- Pandas has a myriad of built-in methods for various computations


- In this lesson, we'll learn about some of the basic ones


- To find the documentation for the Series and DataFrame methods covered in this section, [search the documentation](https://pandas.pydata.org/pandas-docs/stable/index.html) for either
    - `pandas.Series.method_name` or
    - `pandas.DataFrame.method_name`

### Element-wise methods

* **Element-wise methods** perform the same computation on each element of a Series or DataFrame


* One such example is the `.round()` Series/DataFrame method


* When applied to a Series, we get the values in the Series, rounded to the given number of decimals


* For example:

In [11]:
# Solution
df['d'].round(2)

0    14.12
1    15.84
2    15.19
3    15.30
4    14.79
Name: d, dtype: float64

* We can add a column to our DataFrame with these rounded values, like this:

In [12]:
# Solution
df.assign(
    d_rounded=lambda x: x['d'].round(2)
)

a         b   c          d  d_rounded
0  5  8.389033  10  14.115655      14.12
1  3  9.586854  17  15.843354      15.84
2  2  5.110890  17  15.194550      15.19
3  7  7.578774  14  15.297023      15.30
4  4  1.602499  18  14.786111      14.79

* We can also apply `.round()` to the entire DataFrame:

In [13]:
# Solution
df.round(1)

a    b   c     d
0  5  8.4  10  14.1
1  3  9.6  17  15.8
2  2  5.1  17  15.2
3  7  7.6  14  15.3
4  4  1.6  18  14.8

- A few mathematical element-wise methods in Pandas:

| Method | Description |
| :- | :- |
| `.round()` | Round values to given number of decimal places | 
| `.abs()` | Absolute values |


- For more complex element-wise mathematical computations, you can use the appropriate [NumPy](https://numpy.org/) functions


- For example, we can compute the cosine of the values in column `b` like this:

In [14]:
# Solution
# Import NumPy
import numpy as np

# Create a new variable with the cosine of the values in column 'b'
df.assign(
    cos_b=lambda x: np.cos(x['b'])
)

a         b   c          d     cos_b
0  5  8.389033  10  14.115655 -0.509885
1  3  9.586854  17  15.843354 -0.986894
2  2  5.110890  17  15.194550  0.388037
3  7  7.578774  14  15.297023  0.271746
4  4  1.602499  18  14.786111 -0.031698

- [Here's a list of NumPy mathematical functions](https://numpy.org/doc/stable/reference/routines.math.html)

❓ __Exercise 1.__ Add a column to `df` called `log10_c`, which contains the base 10 logarithm of the values in column `c`.

*Hint.* Take a look at the list of NumPy mathematical functions linked above.

In [15]:
# Solution
df.assign(
    log10_c=lambda x: np.log10(x['c'])
)

a         b   c          d   log10_c
0  5  8.389033  10  14.115655  1.000000
1  3  9.586854  17  15.843354  1.230449
2  2  5.110890  17  15.194550  1.230449
3  7  7.578774  14  15.297023  1.146128
4  4  1.602499  18  14.786111  1.255273

### Reduction methods

- **Reduction methods** either
    - map a Series to a single value, or
    - map the rows or columns of a DataFrame to a Series


- One such example is the `.sum()` Series/DataFrame method


- When applied to a Series, we get the sum of the values in the Series:

In [16]:
# Solution
df['b'].sum()

32.268050202

- By default, when applied to a DataFrame, `.sum()` returns a Series containing sums *across the rows*, or within each column:

In [17]:
# Solution
df.sum()
# df.sum(axis='rows') returns the same thing

a    21.000000
b    32.268050
c    76.000000
d    75.236694
dtype: float64

- To sum *across the columns* instead, we can use the `axis='columns'` keyword:

In [18]:
# Solution
df.sum(axis='columns')

0    37.504688
1    45.430208
2    39.305440
3    43.875797
4    38.388611
dtype: float64

- We can select a subset of columns first, and then sum across the columns, like this:

In [19]:
# Solution
df[['a', 'c', 'd']].sum(axis='columns')

0    29.115655
1    35.843354
2    34.194550
3    36.297023
4    36.786111
dtype: float64

* We can add a column to our DataFrame with this sum, like this:

In [20]:
# Solution
df.assign(
    total_no_b=lambda x: x[['a', 'c', 'd']].sum(axis='columns')
)

a         b   c          d  total_no_b
0  5  8.389033  10  14.115655   29.115655
1  3  9.586854  17  15.843354   35.843354
2  2  5.110890  17  15.194550   34.194550
3  7  7.578774  14  15.297023   36.297023
4  4  1.602499  18  14.786111   36.786111

*Food for thought.* What happens when you use the code above, but with `axis='rows'` instead? Why does this happen?

- Here are some useful reduction methods:

    
| Method | Description |
| :- | :- |
| `.count()` | Number of non-NA values |
| `.min()`, `.max()` | Compute minimum and maximum values |
| `.quantile()` | Compute sample quantile ranging from 0 to 1 |
| `.sum()` | Sum of values |
| `.mean()` | Mean of values |
| `.median()` | Median of values |
| `.mad()` | Mean absolute deviation from mean value |
| `.prod()` | Product of all values |
| `.var()` | Sample variance of values |
| `.std()` | Sample standard deivation of values |


- All the methods above use the `axis=...` keyword argument in the same way

❓ __Exercise 2.__ Add two new columns to `df` called `min_value` and `max_value`, containing the minimum and maximum value in each row, respectively.

In [21]:
# Solution
df.assign(
    min_value=lambda x: x.min(axis='columns'),
    max_value=lambda x: x.max(axis='columns')
)

a         b   c          d  min_value  max_value
0  5  8.389033  10  14.115655   5.000000  14.115655
1  3  9.586854  17  15.843354   3.000000  17.000000
2  2  5.110890  17  15.194550   2.000000  17.000000
3  7  7.578774  14  15.297023   7.000000  15.297023
4  4  1.602499  18  14.786111   1.602499  18.000000

### "Same size" methods

* There are other methods that
    - take a row or column of values
    - perform computations on these values *as a group* (not element-wise)
    - return a row or column of values *of the same size*


* For example, we can obtain the cumulative sum of values across the *columns* with the `.cumsum()` method, like this:

In [22]:
# Solution
df.cumsum(axis='columns')

a          b          c          d
0  5.0  13.389033  23.389033  37.504688
1  3.0  12.586854  29.586854  45.430208
2  2.0   7.110890  24.110890  39.305440
3  7.0  14.578774  28.578774  43.875797
4  4.0   5.602499  23.602499  38.388611

* We can also obtain the cumulative sum of values across the *rows* of a column with the `.cumsum()` method, like this:

In [23]:
# Solution
df['c'].cumsum(axis='rows')
# df['c'].cumsum() returns the same thing

0    10
1    27
2    44
3    58
4    76
Name: c, dtype: int64

* As we have done above, we can add this cumulative sum of values for column `c` as a new column of our DataFrame, like this:

In [24]:
# Solution
df.assign(
    c_cumsum=lambda x: x['c'].cumsum(axis='rows')
)

a         b   c          d  c_cumsum
0  5  8.389033  10  14.115655        10
1  3  9.586854  17  15.843354        27
2  2  5.110890  17  15.194550        44
3  7  7.578774  14  15.297023        58
4  4  1.602499  18  14.786111        76

* Here are some methods that behave similarly &mdash; that is, they
    - take a row or column of values
    - perform a computation on these values as a group
    - then return a row or column of values of the same size

| Method | Description |
| :- | :- |
| `.cumsum()` | Cumulative sum of values |
| `.cummin()`, `.cummax()` | Cumulative minimum and maximum of values |
| `.cumprod()` | Cumulative product of values |
| `.diff()` | Compute difference between consecutive rows/columns |
| `.pct_change()` | Compute percent changes between consecutive rows/columns |
| `.rank()` | Compute numerical ranks | 


* Some of these methods take optional keyword arguments that change what they compute 
    * Take a look at the documentation for details

❓ __Exercise 3.__ Add a column called `a_diff` to `df`, containing the difference between consecutive values in column `a`. 

You should see that the first element of your newly created column `a_diff` has a missing value. Does this make sense?

In [25]:
# Solution
df.assign(
    a_diff=lambda x: x['a'].diff()
)

a         b   c          d  a_diff
0  5  8.389033  10  14.115655     NaN
1  3  9.586854  17  15.843354    -2.0
2  2  5.110890  17  15.194550    -1.0
3  7  7.578774  14  15.297023     5.0
4  4  1.602499  18  14.786111    -3.0

*Write your answer here. Double-click to edit.*

*Solution.* We expect that the first element of `a_diff` is missing, since the first row does not have a previous element to calculate a difference with.

<hr style="border-top: 2px solid gray; margin-top: 1px; margin-bottom: 1px"></hr>

## Other useful ways to create new variables

### Adding a constant column

* Sometimes it's useful to add a column in which every value is equal to a constant, either a string or numeric value


* In this case, you don't need to use a lambda function, like this:

In [26]:
# Solution
df.assign(
    year=1999, month='August'
)

a         b   c          d  year   month
0  5  8.389033  10  14.115655  1999  August
1  3  9.586854  17  15.843354  1999  August
2  2  5.110890  17  15.194550  1999  August
3  7  7.578774  14  15.297023  1999  August
4  4  1.602499  18  14.786111  1999  August

### Mapping values in one column to another

* Suppose we know that the values in column `c` in `df` range between 10 and 20, and we want to create a column with those values in words

* We can create a dictionary with our desired mapping, like this:

In [27]:
number_in_words = {
    10: 'Ten',
    11: 'Eleven',
    12: 'Twelve',
    13: 'Thirteen',
    14: 'Fourteen',
    15: 'Fifteen',
    16: 'Sixteen',
    17: 'Seventeen',
    18: 'Eighteen',
    19: 'Nineteen',
    20: 'Twenty'
}

* Then, we can use the `.map()` method to create our new column, like this:

In [28]:
# Solution
df.assign(
    c_words=lambda x: x['c'].map(number_in_words)
)

a         b   c          d    c_words
0  5  8.389033  10  14.115655        Ten
1  3  9.586854  17  15.843354  Seventeen
2  2  5.110890  17  15.194550  Seventeen
3  7  7.578774  14  15.297023   Fourteen
4  4  1.602499  18  14.786111   Eighteen

* We can also pass a _function_ to `.map()` to accomplish something similar!

* For example, suppose we want to create a new column based on column `b` in `df`:
    * If the value in `b` is below 6, we want the value in our new column to be `low`
    * Otherwise, we want the value in our new column to be `high`
    
* Let's first define a function that creates this mapping:

In [29]:
def low_or_high(value):
    if value < 6:
        return 'low'
    else:
        return 'high'

* Now we can use the `.map()` method to create our new column, like this:

In [30]:
# Solution
df.assign(
    b_level=lambda x: x['b'].map(low_or_high)
)

a         b   c          d b_level
0  5  8.389033  10  14.115655    high
1  3  9.586854  17  15.843354    high
2  2  5.110890  17  15.194550     low
3  7  7.578774  14  15.297023    high
4  4  1.602499  18  14.786111     low

<hr style="border-top: 2px solid gray; margin-top: 1px; margin-bottom: 1px"></hr>

## Problems

### Problem 0

In the same folder as this notebook, there is a CSV file `data/grades.csv`. This is the same dataset used for the problems in Lesson 12. Read the CSV file into a DataFrame. Display the top 5 rows of the DataFrame.

In [31]:
# Solution
grades_df = pd.read_csv('data/grades.csv')
grades_df.head()

student_id    lastname firstname  quiz1  quiz2  quiz3  quiz4  exam1  exam2
0        2603      Conway     Romeo     19     29     20     30     87     50
1        6435  Livingston    Naseem     17     16     14     25     55     83
2        6754       Clark      Remy     15     30     11     24     60     95
3        3032   Carpenter    Zubair     16     13     18     14     91     71
4        2715      Guerra  Samantha     13     15     17     28     48     98

### Problem 1

Find the minimum, maximum, and median quiz and exam grades. Write your code so that the student IDs and names are *not* included in the output.

*Hint.* Use the `.drop()` DataFrame method.

In [32]:
# Solution
grades_df.drop(
    columns=['student_id', 'lastname', 'firstname']
).min()

quiz1    10
quiz2    11
quiz3    10
quiz4    10
exam1    41
exam2    44
dtype: int64

In [33]:
# Solution
grades_df.drop(
    columns=['student_id', 'lastname', 'firstname']
).max()

quiz1    20
quiz2    30
quiz3    20
quiz4    30
exam1    99
exam2    99
dtype: int64

In [34]:
# Solution
grades_df.drop(
    columns=['student_id', 'lastname', 'firstname']
).median()

quiz1    14.0
quiz2    20.0
quiz3    16.0
quiz4    19.0
exam1    72.0
exam2    73.0
dtype: float64

### Problem 2

Create a new DataFrame based on the one you created in Problem 0:

- Compute each student's quiz grades as a percentage (fraction between 0 and 1). The maximum score on Quizzes 1 and 3 is 20 points; the maximum score on Quizzes 2 and 4 is 30 points. Add 4 columns, one for each quiz as a percentage.


- Compute each student's exam grades as a percentage. The maximum score on each exam is 100 points. Add 2 columns, one for each exam as a percentage.


- For each student, compute the average of their quiz percentage grades. Add a column containing the quiz averages.


- For each student, compute the average of their exam percentage grades. Add a column containing the exam averages.


- Compute each student's course grade as a weighted average of each student's quiz average and exam average: quizzes are worth 35\%, exams are worth 65\%. Add a column containing the course grade.


Do this using a *single* call to `.assign()`, and use as many of the Series/DataFrame methods described above (e.g. `.sum()`, `.mean()`, etc.) as required. 

Check your work by displaying the top 5 rows of the DataFrame. You should find that Romeo Conway has a course grade of about 78.8%, Naseem Livingston has a course grade of 70.4%, and Remy Clark has a course grade of 77.5%.

In [35]:
# Solution
grades_df = grades_df.assign(
    quiz1_pct=lambda x: x['quiz1'] / 20,
    quiz2_pct=lambda x: x['quiz2'] / 30,
    quiz3_pct=lambda x: x['quiz3'] / 20,
    quiz4_pct=lambda x: x['quiz4'] / 30,
    exam1_pct=lambda x: x['exam1'] / 100,
    exam2_pct=lambda x: x['exam2'] / 100,
    quiz_avg=lambda x: x[['quiz1_pct', 'quiz2_pct', 'quiz3_pct', 'quiz4_pct']].mean(axis='columns'),
    exam_avg=lambda x: x[['exam1_pct', 'exam2_pct']].mean(axis='columns'),
    course_grade=lambda x: 0.35 * x['quiz_avg'] + 0.65 * x['exam_avg']
)

grades_df.head()

student_id    lastname firstname  quiz1  quiz2  quiz3  quiz4  exam1  exam2  \
0        2603      Conway     Romeo     19     29     20     30     87     50   
1        6435  Livingston    Naseem     17     16     14     25     55     83   
2        6754       Clark      Remy     15     30     11     24     60     95   
3        3032   Carpenter    Zubair     16     13     18     14     91     71   
4        2715      Guerra  Samantha     13     15     17     28     48     98   

   quiz1_pct  quiz2_pct  quiz3_pct  quiz4_pct  exam1_pct  exam2_pct  quiz_avg  \
0       0.95   0.966667       1.00   1.000000       0.87       0.50  0.979167   
1       0.85   0.533333       0.70   0.833333       0.55       0.83  0.729167   
2       0.75   1.000000       0.55   0.800000       0.60       0.95  0.775000   
3       0.80   0.433333       0.90   0.466667       0.91       0.71  0.650000   
4       0.65   0.500000       0.85   0.933333       0.48       0.98  0.733333   

   exam_avg  course_grade  
0     0.685      0.787958  
1     0.690      0.703708  
2     0.775      0.775000  
3     0.810      0.754000  
4     0.730      0.731167

### Problem 3

Create a new DataFrame based on the one you created in Problem 2. In particular, add a new column containing the rank of each student based on their course grade. The student with the highest course grade should have a rank 1. Your new DataFrame should *only* contain the student ID, last name and first name, course grade, and rank. The first 5 rows of your new DataFrame should look like this:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>student_id</th>
      <th>lastname</th>
      <th>firstname</th>
      <th>course_grade</th>
      <th>rank</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2603</td>
      <td>Conway</td>
      <td>Romeo</td>
      <td>0.787958</td>
      <td>8.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>6435</td>
      <td>Livingston</td>
      <td>Naseem</td>
      <td>0.703708</td>
      <td>30.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>6754</td>
      <td>Clark</td>
      <td>Remy</td>
      <td>0.775000</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3032</td>
      <td>Carpenter</td>
      <td>Zubair</td>
      <td>0.754000</td>
      <td>14.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2715</td>
      <td>Guerra</td>
      <td>Samantha</td>
      <td>0.731167</td>
      <td>19.0</td>
    </tr>
  </tbody>
</table>

*Hint.* Read the documentation for `.rank()`.

In [36]:
# Solution
ranked_df = (
    grades_df
    .assign(
        rank=lambda x: x['course_grade'].rank(ascending=False)
    )
    [['student_id', 'lastname', 'firstname', 'course_grade', 'rank']]
)

ranked_df.head()

student_id    lastname firstname  course_grade  rank
0        2603      Conway     Romeo      0.787958   8.0
1        6435  Livingston    Naseem      0.703708  30.0
2        6754       Clark      Remy      0.775000  10.0
3        3032   Carpenter    Zubair      0.754000  14.0
4        2715      Guerra  Samantha      0.731167  19.0